# **Homework Assignment #5**

Assigned: March 23, 2022

Due: April 11, 2022

Name: Brandon Luy
ID: 11532759

---

1. ![](https://drive.google.com/uc?export=view&id=1p9q7eJIZGh1EbM3TIE45dgg6gfrM_f0x)


2. (10 points) In class, we discussed regularization terms that could be added to a perceptron loss function to control its performance. Describe ways that regularization can be applied to these other types of learning algorithms:

*   Decision Tree
    *   Regularization can be applied onto decision trees to prevent overfitting could be by limiting the maximum depth of trees, and setting more specfic needs on whne to keep splitting a node further such as the number of samples.

*   Support Vector Machine
    *   Regularization can be applied to Support Vector Machines by as a parameter, where for larger values of C (reg. parameter) it chooses a smaller margin hyperplane and for smaller values of C, it looks for larger margin seperating hyperplane. 

*   Naive Bayes classifier
    *   Regularization in Naive Bayes can be applied in 2 ways where l1 is applied to shrink the parameters, selects varibales, dropping irrelevant and redundant variables. Whereas, l2 just shrinks the parameters.



3. 
(20 points) Describe the advantages and disadvantages of using ReLU as an activation function instead of tanh. Additionally, describe one change to ReLU that could be introduced to mitigate the disadvantage that you describe.


*  Advantages: One advantage for using ReLU would be that it saves proccessing power because it implements the max() function compared to an exponential function. Also, it is able to output a true zero value, and acts like a linear function.
*  Disadvantages: Since ReLU is only when the units are positive and zeros, the problems that can arise from this can be leftover positve bias for later layers because the mean activiation is greater than zero. One way to mitigate this disadvantage can be using batch normalization to move the mean towards zero.




4. 
Consider a two-layer neural network containing three hidden units. All values of  𝑊  are 1.0 and  𝑣1=−0.4 ,  𝑣2=1.0 ,  𝑣3=0.2 . For X =  < 0, 1, 1, 0 > , show each step of how the vector of weights  𝑣  will be updated using backpropagation and  𝜂=0.5 , given that the actual value of  𝑦  is 0.1.
    *   ![](https://drive.google.com/uc?export=view&id=1UrXKPHVz1N1nE10zkh4uaXtnUObD0fUn)



5.
(80 points) In this programming assignment, you will run and modify a simple recurrent neural network (RNN) that predicts the next word in a nursery rhythm.
𝐵𝑎𝑐𝑘𝑔𝑟𝑜𝑢𝑛𝑑.  Below is code for an RNN that predicts the next word in the "Twinkle Twinkle" nursery rhyme. A RNN processes sequential data. Instead of feeding all available information into a neural network, a RNN feeds information from one point in the sequence at a time. However, the RNN contains a loop so that the output from one point in the sequence feeds back into the network to capture context that helps in prediction.
This particular RNN is a special type of structure called a long short-term memory network (LSTM). An LSTM stores extra information and can add or remove information from a cell state. This allows the neural network to learn more complex patterns. This example code creates a sequence of random integers for training and testing.
𝑌𝑜𝑢𝑟   𝑐𝑜𝑛𝑡𝑟𝑖𝑏𝑢𝑡𝑖𝑜𝑛.  Please complete the following tasks:
1) Compare the accuracy-during-training performance for two other context sizes (number of words used as the feature vector). Which performed best for these data?
2) Test the model using the second verse of the nursery rhyme (found at https://parenting.firstcry.com/articles/10-best-nursery-rhymes-for-babies/). How does the performance compare with that of training for the three chosen context sizes?
3) Expand this text prediction code to read in a larger document (one of the newsgroups postings or an entire newsgroup from the 20 newsgroup example we used in class and available from sklearn is one possibility but you can select another). Report performance for training and testing from subsequences of this text. You might want to employ additional text processing features we highlighted in class to further clean and preprocess the text as needed.

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras import layers

data = """ Twinkle, twinkle, little star
            How I wonder what you are \
            Up above the world so high \
            Like a diamond in the sky \
            Twinkle, twinkle little star \
            How I wonder what you are\n """
seq_size = 3

tokenizer = Tokenizer()
tokenizer.fit_on_texts([data])
sequence = tokenizer.texts_to_sequences([data])[0]

X = []
y = []

# input is elements i-seq_size through i-1, output is element i
for i in range(seq_size, len(sequence)):
    X.append(sequence[i-seq_size:i])
    y.append(sequence[i])
X = np.array(X)
y = np.array(y)
X = X.reshape((X.shape[0], X.shape[1], 1))

model = tf.keras.Sequential()
model.add(layers.LSTM(50, activation='relu', input_shape=(seq_size, 1)))
model.add(layers.Dense(1))

model.compile(optimizer=tf.keras.optimizers.Adam(0.01),
              loss=tf.keras.losses.MeanSquaredError(), metrics=['accuracy'])
model.fit(X, y, epochs=200, verbose=1)
score = model.evaluate(X, y , verbose=1)
print('Test Accuracy: ', score)


Epoch 1/200
1/1 [==============================] - 1s 1s/step - loss: 122.6607 - accuracy: 0.0000e+00
Epoch 2/200
1/1 [==============================] - 0s 11ms/step - loss: 102.6371 - accuracy: 0.0690
Epoch 3/200
1/1 [==============================] - 0s 11ms/step - loss: 81.9649 - accuracy: 0.0690
Epoch 4/200
1/1 [==============================] - 0s 9ms/step - loss: 59.9946 - accuracy: 0.0690
Epoch 5/200
1/1 [==============================] - 0s 10ms/step - loss: 38.8558 - accuracy: 0.0690
Epoch 6/200
1/1 [==============================] - 0s 14ms/step - loss: 26.0186 - accuracy: 0.0690
Epoch 7/200
1/1 [==============================] - 0s 14ms/step - loss: 36.0625 - accuracy: 0.0690
Epoch 8/200
1/1 [==============================] - 0s 9ms/step - loss: 39.6728 - accuracy: 0.0690
Epoch 9/200
1/1 [==============================] - 0s 14ms/step - loss: 32.1395 - accuracy: 0.0690
Epoch 10/200
1/1 [==============================] - 0s 10ms/step - loss: 25.2939 - accuracy: 0.0690
Epoch 1

In [ ]:
#Part 2
import numpy as np
import tensorflow as tf
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras import layers

data = """ When the blazing sun is gone
            When he nothing shines upon \
            Then you show your little light \
            Twinkle, twinkle, all the night \
            Twinkle, twinkle, little star \
            How I wonder what you are  """
seq_size = 3

tokenizer = Tokenizer()
tokenizer.fit_on_texts([data])
sequence = tokenizer.texts_to_sequences([data])[0]

X = []
y = []

# input is elements i-seq_size through i-1, output is element i
for i in range(seq_size, len(sequence)):
    X.append(sequence[i-seq_size:i])
    y.append(sequence[i])
X = np.array(X)
y = np.array(y)
X = X.reshape((X.shape[0], X.shape[1], 1))

model = tf.keras.Sequential()
model.add(layers.LSTM(50, activation='relu', input_shape=(seq_size, 1)))
model.add(layers.Dense(1))

model.compile(optimizer=tf.keras.optimizers.Adam(0.01),
              loss=tf.keras.losses.MeanSquaredError(), metrics=['accuracy'])
model.fit(X, y, epochs=200, verbose=1)
score = model.evaluate(X, y , verbose=1)
print('Test Accuracy: ', score)


Epoch 1/200
1/1 [==============================] - 1s 1s/step - loss: 186.1186 - accuracy: 0.0690
Epoch 2/200
1/1 [==============================] - 0s 9ms/step - loss: 162.0331 - accuracy: 0.0690
Epoch 3/200
1/1 [==============================] - 0s 13ms/step - loss: 140.5014 - accuracy: 0.1379
Epoch 4/200
1/1 [==============================] - 0s 9ms/step - loss: 116.9497 - accuracy: 0.1379
Epoch 5/200
1/1 [==============================] - 0s 9ms/step - loss: 91.3247 - accuracy: 0.1379
Epoch 6/200
1/1 [==============================] - 0s 8ms/step - loss: 74.9717 - accuracy: 0.1379
Epoch 7/200
1/1 [==============================] - 0s 7ms/step - loss: 91.6158 - accuracy: 0.1379
Epoch 8/200
1/1 [==============================] - 0s 15ms/step - loss: 90.8974 - accuracy: 0.1379
Epoch 9/200
1/1 [==============================] - 0s 15ms/step - loss: 78.3362 - accuracy: 0.1379
Epoch 10/200
1/1 [==============================] - 0s 10ms/step - loss: 70.3791 - accuracy: 0.1379
Epoch 11/200

The perfomance is much different compared to the second verse, where here in the second verse the Loss was reported as 3.449, and Accuracy at 0.0690. Compared to in the first verse where: [0.2868848443031311, 0.06896551698446274]

---


#Project Proposal.

(10 points, counts toward semester project) Provide a brief description (1-2 paragraphs) of your proposed project. The description should include the following:

* Project statement. This includes the algorithm you will implement, modify, or apply; the method(s) for data preparation, learning, visualization you will consider; the application that will use the results, if appropriate; the evaluation mechanisms you will include.

* Team roles. Team projects are encouraged, although not required. Team sizes should be 2-3 members, any larger team needs special permission. Clearly delinate the role of each team member in the project.

* Timeline for accomplishing the project goals.



*   Project Statement: 
  The project that I will be implemeting will include a the main implementation of Classification to identify whether a photo contains a photo of a human, dog, or cat. This implementation will include a recently used topic to implement CNN which is primarily used to classify images, and identify patterns between them. The box of the picture is intepreteded by the number of pixels defining the height and width of the box, and with a depth of 3 for each letter pertaining to RGB. Then, pass each square of pixels and pass it through a filter.
    As of right now, the  main methods that will be used for data processing will be the python library pandas, pyplot from matlib for an intial viewing of feedback and results, and tensorflow for image classification from the trained data, finding similarities. 


*   Team Roles: The memebers in the team just consist of me (Brandon Luy). 

